R/get_datagrid.R

```diff
@@ -908,9 +657,287 @@ get_datagrid.datagrid <- get_datagrid.visualisation_matrix
 
 
 
+# Functions for emmeans/marginaleffects ---------------
 
-# helper -----------------
+#' @rdname get_datagrid
+#' @export
+get_datagrid.emmGrid <- function(x, ...) {
+  suppressWarnings(s <- as.data.frame(x))
+  # We want all the columns *before* the estimate column
+  est_col_idx <- which(colnames(s) == attr(s, "estName"))
+  which_cols <- seq_len(est_col_idx - 1)
+
+  data.frame(s)[, which_cols, drop = FALSE]
+}
+
+#' @export
+get_datagrid.emm_list <- function(x, ...) {
+  k <- length(x)
+  res <- vector("list", length = k)
+  for (i in seq_len(k)) {
+    res[[i]] <- get_datagrid(x[[i]])
+  }
 
+  all_cols <- Reduce(lapply(res, colnames), f = union)
+  for (i in seq_len(k)) {
+    res[[i]][,setdiff(all_cols, colnames(res[[i]]))] <- NA
+  }
+  out <- do.call("rbind", res)
+
+  clear_cols <- colnames(out)[sapply(out, Negate(anyNA))] # these should be first
+  out[,c(clear_cols, setdiff(colnames(out), clear_cols)), drop = FALSE]
+}
+
+#' @rdname get_datagrid
+#' @export
+get_datagrid.slopes <- function(x, ...) {
+  cols_newdata <- colnames(attr(x, "newdata"))
+  cols_contrast <- colnames(x)[grep("^contrast_?", colnames(x))]
```

Comment:
I think you can use `startsWith()` instead of grep here, which should be faster.

``` r
microbenchmark::microbenchmark(
  grepl("^Species?", colnames(iris)),
  startsWith(colnames(iris), "Species"),
  times = 1000
)
#> Unit: microseconds
#>                                   expr   min     lq      mean median     uq
#>     grepl("^Species?", colnames(iris)) 9.001 10.202 14.314188 14.001 16.701
#>  startsWith(colnames(iris), "Species") 1.900  2.201  3.807115  2.501  3.901
#>      max neval cld
#>   81.401  1000  a 
#>  957.100  1000   b
```

<sup>Created on 2024-09-03 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

Reply:
But I need that optional `_` character :/

Reply:
Ok.